consensus/src/chained_bft/block_storage/block_tree.rs

```diff
@@ -249,23 +264,42 @@ where
         if let Some(old_qc) = self.id_to_quorum_cert.get(&block_id) {
             return VoteReceptionResult::OldQuorumCertificate(Arc::clone(old_qc));
         }
-        // All the votes collected for all the execution results of a given proposal.
-        let block_votes = self
-            .id_to_votes
-            .entry(block_id)
-            .or_insert_with(HashMap::new);
+
+        let author = vote_msg.author();
+        if let Some(old_block_id) = self.author_to_block.get(&author) {
+            if block_id == *old_block_id {
+                // Author has already voted for this block
+                return VoteReceptionResult::DuplicateVote;
+            }
+            if let Some(old_block_hash) = self.id_to_votes.get_mut(&old_block_id) {
+                let digest = old_block_hash.digest;
+                let old_block_votes = &mut old_block_hash.li_sig;
+                if let Some(li_sig) = old_block_votes.get_mut(&digest) {
+                    if li_sig.remove_signature(author) {
+                        // Last vote/signature for block removed, cleanup hashmap.
+                        old_block_votes.remove(&digest);
+                        self.id_to_votes.remove(&old_block_id);
+                    }
+                }
+            }
+            self.author_to_block.remove(&author);
+        }
+        self.author_to_block.entry(author).or_insert(block_id);
```

Comment:
You don't really care about the previous entry here, calling `insert` seems to make more sense.